gtsam/linear/PowerMethod.h

```diff
@@ -0,0 +1,139 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010-2019, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file   PowerMethod.h
+ * @date   Sept 2020
+ * @author Jing Wu
+ * @brief  Power method for fast eigenvalue and eigenvector
+ * computation
+ */
+
+#pragma once
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/Vector.h>
+
+#include <Eigen/Core>
+#include <Eigen/Sparse>
+#include <random>
+#include <vector>
+
+namespace gtsam {
+
+using Sparse = Eigen::SparseMatrix<double>;
+
+/* ************************************************************************* */
+/// MINIMUM EIGENVALUE COMPUTATIONS
+
+// Template argument Operator just needs multiplication operator
+template <class Operator>
+class PowerMethod {
+ protected:
+  // Const reference to an externally-held matrix whose minimum-eigenvalue we
+  // want to compute
+  const Operator &A_;
+
+  const int dim_;  // dimension of Matrix A
+
+  size_t nrIterations_;  // number of iterations
+
+  double ritzValue_;   // all Ritz eigenvalues
+  Vector ritzVector_;  // all Ritz eigenvectors
+
+ public:
+  // Constructor
+  explicit PowerMethod(const Operator &A,
+                       const boost::optional<Vector> initial = boost::none)
+      : A_(A), dim_(A.rows()), nrIterations_(0) {
+    Vector x0;
+    x0 = initial ? Vector::Random(dim_) : initial.get();
+    x0.normalize();
+
+    // initialize Ritz eigen values
+    ritzValue_ = 0.0;
+
+    // initialize Ritz eigen vectors
+    ritzVector_ = Vector::Zero(dim_);
+
+    ritzVector_.col(0) = update(x0);
+    perturb();
+  }
+
+  // Update the vector by dot product with A_
+  Vector update(const Vector &x) const {
+    Vector y = A_ * x;
+    y.normalize();
+    return y;
+  }
+
+  // Update the vector by dot product with A_
+  Vector update() const { return update(ritzVector_); }
+
+  // Perturb the initial ritzvector
```

Comment:
This is not well documented. Where does this come from?

Reply:
It is now moved to ShonanAveraging.cpp, it comes from Part III.C David's paper Rosen, D. and Carlone, L., 2017, September. Computational enhancements for certifiably correct SLAM. Already added this in the documentation.